File: elixir/lib/rss_fetcher.ex
# Tanya RSS Fetcher in Elixir
# Run: mix run elixir/rss_fetcher.exs
# Requires: Mix project setup

defmodule Tanya.RSSFetcher do
  @moduledoc """
  RSS Feed Fetcher - Elixir implementation
  """

  @feeds [
    %{name: "BBC", url: "http://feeds.bbci.co.uk/news/world/rss.xml", category: "World"},
    %{name: "TechCrunch", url: "https://techcrunch.com/feed/", category: "Tech"},
  ]

  @doc """
  Fetch RSS from URL
  """
  def fetch(url) do
    case HTTPoison.get(url, [{"User-Agent", "Tanya/1.0 (Elixir)"}]) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
        {:ok, body}
      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Parse RSS XML
  """
  def parse(xml) do
    items = []
    
    # Extract titles (simplified)
    titles = Regex.scan(~r/<title>([^<]+)<\/title>/, xml, capture: :all)
              |> Enum.map(fn [_, title] -> title end)
              |> Enum.drop(1)  # Drop channel title
    
    Enum.map(titles, fn title ->
      %{
        title: title,
        sentiment: analyze_sentiment(title),
        category: infer_category(title),
        source: "Elixir"
      }
    end)
  end

  @doc """
  Simple sentiment analysis
  """
  def analyze_sentiment(text) do
    text = String.downcase(text)
    
    positive = ~w(good great excellent amazing breakthrough success win)
    negative = ~w(bad terrible crisis fail death war attack disaster)
    
    pos_count = Enum.count(positive, &String.contains?(text, &1))
    neg_count = Enum.count(negative, &String.contains?(text, &1))
    
    cond do
      pos_count > neg_count -> "positive"
      neg_count > pos_count -> "negative"
      true -> "neutral"
    end
  end

  @doc """
  Infer category from title
  """
  def infer_category(title) do
    text = String.downcase(title)
    
    cond do
      String.contains?(text, "ai") or String.contains?(text, "artificial intelligence") -> "AI"
      String.contains?(text, "tech") or String.contains?(text, "software") -> "Tech"
      String.contains?(text, "stock") or String.contains?(text, "market") -> "Finance"
      String.contains?(text, "war") or String.contains?(text, "military") -> "World"
      String.contains?(text, "science") or String.contains?(text, "space") -> "Science"
      true -> "General"
    end
  end

  @doc """
  Main runner
  """
  def run(args \\ []) do
    IO.puts("Tanya RSS Fetcher (Elixir)")
    IO.puts("===========================")
    IO.puts("")

    feeds = if "--url" in args do
      idx = Enum.find_index(args, &(&1 == "--url"))
      if idx && Enum.at(args, idx + 1) do
        [%{name: "Custom", url: Enum.at(args, idx + 1), category: "General"}]
      else
        @feeds
      end
    else
      @feeds
    end

    all_items = []
    
    for feed <- feeds do
      IO.puts("Fetching #{feed.name}...")
      
      case fetch(feed.url) do
        {:ok, xml} ->
          items = parse(xml)
          IO.puts("  -> Found #{length(items)} articles")
          Enum.each(items, fn item ->
            IO.puts("    - #{item.title}")
          end)
          all_items = items ++ all_items
          
        {:error, reason} ->
          IO.puts("  -> Error: #{reason}")
      end
    end
    
    IO.puts("")
    IO.puts("Total: #{length(all_items)} articles")
    IO.puts("Engine: Elixir (fault-tolerant, distributed)")
  end
end

# If run directly
if System.get_env("MIX_ENV") == nil || true do
  Tanya.RSSFetcher.run(System.argv())
end

File: rust/src/rss_fetcher.rs
//! RSS Fetcher - Fetches and parses RSS/Atom feeds
//! Build: cd rust && cargo build --release
//! Run:   ./target/release/rss_fetcher <feed_url>

use rss::{Channel, ChannelBuilder, Item, ItemBuilder};
use serde::{Deserialize, Serialize};
use std::env;
use std::fs;
use std::io::Write;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct NewsItem {
    pub title: String,
    pub link: String,
    pub description: String,
    pub pub_date: String,
    pub source: String,
    pub category: String,
    pub reading_time: u32,
    pub sentiment: String,
    pub keywords: Vec<String>,
}

fn estimate_reading_time(content: &str) -> u32 {
    let words: Vec<&str> = content.split_whitespace().collect();
    let minutes = (words.len() as f64 / 200.0).ceil() as u32;
    minutes.max(1)
}

fn extract_keywords(title: &str, desc: &str) -> Vec<String> {
    let text = format!("{} {}", title, desc);
    let stop_words = ["the", "a", "an", "and", "or", "but", "in", "on", "at", "to", "for", 
                      "of", "with", "by", "from", "is", "are", "was", "were", "be", "been",
                      "being", "have", "has", "had", "do", "does", "did", "will", "would",
                      "could", "should", "may", "might", "must", "shall", "can", "need",
                      "dare", "ought", "used", "this", "that", "these", "those", "it", "its"];
    
    let words: Vec<String> = text.to_lowercase()
        .split(|c: char| !c.is_alphanumeric())
        .filter(|w| w.len() > 3)
        .filter(|w| !stop_words.contains(w))
        .map(|w| w.to_string())
        .collect();
    
    let mut word_count: std::collections::HashMap<String, u32> = std::collections::HashMap::new();
    for word in words {
        *word_count.entry(word).or_insert(0) += 1;
    }
    
    let mut keywords: Vec<String> = word_count.into_iter()
        .filter(|(_, count)| *count > 1)
        .map(|(word, _)| word)
        .take(10)
        .collect();
    
    keywords.sort();
    keywords
}

fn analyze_sentiment(title: &str, desc: &str) -> String {
    let text = format!("{} {}", title, desc).to_lowercase();
    
    let positive = ["good", "great", "excellent", "amazing", "breakthrough", "success", 
                    "win", "achievement", "improve", "positive", "growth", "hero", "best"];
    let negative = ["bad", "terrible", "crisis", "fail", "death", "war", "attack", "disaster",
                    "loss", "fear", "worse", "worst", "kill", "destroy", "threat"];
    
    let pos_count = positive.iter().filter(|w| text.contains(*w)).count();
    let neg_count = negative.iter().filter(|w| text.contains(*w)).count();
    
    if pos_count > neg_count {
        "positive".to_string()
    } else if neg_count > pos_count {
        "negative".to_string()
    } else {
        "neutral".to_string()
    }
}

fn infer_category(title: &str, desc: &str) -> String {
    let text = format!("{} {}", title, desc).to_lowercase();
    
    if text.contains("ai") || text.contains("artificial intelligence") || text.contains("machine learning") || text.contains("chatgpt") {
        "AI".to_string()
    } else if text.contains("tech") || text.contains("software") || text.contains("google") || text.contains("microsoft") || text.contains("apple") {
        "Tech".to_string()
    } else if text.contains("stock") || text.contains("market") || text.contains("economy") || text.contains("bitcoin") || text.contains("crypto") {
        "Finance".to_string()
    } else if text.contains("climate") || text.contains("weather") || text.contains("environment") || text.contains("storm") {
        "Weather".to_string()
    } else if text.contains("war") || text.contains("military") || text.contains("russia") || text.contains("ukraine") || text.contains("iran") {
        "World".to_string()
    } else if text.contains("science") || text.contains("space") || text.contains("nasa") || text.contains("research") {
        "Science".to_string()
    } else {
        "General".to_string()
    }
}

fn fetch_feed(url: &str) -> Result<Channel, String> {
    let client = reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(30))
        .build()
        .map_err(|e| e.to_string())?;
    
    let response = client.get(url)
        .header("User-Agent", "Tanya/1.0 RSS Reader")
        .send()
        .map_err(|e| e.to_string())?;
    
    let content = response.bytes().map_err(|e| e.to_string())?;
    Channel::read_from(&content[..]).map_err(|e| e.to_string())
}

fn main() {
    let args: Vec<String> = env::args().collect();
    
    if args.len() < 2 {
        // List sources from JSON
        let sources = load_sources();
        for source in sources {
            match fetch_feed(&source.url) {
                Ok(channel) => {
                    for item in channel.items() {
                        let news_item = process_item(item, &source.name);
                        println!("{}", serde_json::to_string(&news_item).unwrap_or_default());
                    }
                }
                Err(e) => {
                    eprintln!("Error fetching {}: {}", source.name, e);
                }
            }
        }
    } else if args[1] == "--help" {
        println!("Tanya RSS Fetcher");
        println!("Usage:");
        println!("  rss_fetcher              # Fetch all configured sources");
        println!("  rss_fetcher <url>        # Fetch single feed");
        println!("  rss_fetcher --list       # List configured sources");
    } else if args[1] == "--list" {
        let sources = load_sources();
        for source in sources {
            println!("{}: {}", source.name, source.url);
        }
    } else {
        // Fetch single URL
        match fetch_feed(&args[1]) {
            Ok(channel) => {
                for item in channel.items() {
                    let news_item = process_item(item, &channel.title().unwrap_or("Unknown").to_string());
                    println!("{}", serde_json::to_string(&news_item).unwrap_or_default());
                }
            }
            Err(e) => {
                eprintln!("Error: {}", e);
                std::process::exit(1);
            }
        }
    }
}

fn load_sources() -> Vec<FeedSource> {
    let sources_file = std::path::Path::new("../data/sources.json");
    if let Ok(content) = fs::read_to_string(sources_file) {
        if let Ok(sources) = serde_json::from_str::<Vec<FeedSource>>(&content) {
            return sources;
        }
    }
    
    // Default sources
    vec![
        FeedSource { name: "BBC".to_string(), url: "http://feeds.bbci.co.uk/news/world/rss.xml".to_string(), category: "World".to_string() },
        FeedSource { name: "Reuters".to_string(), url: "https://www.reutersagency.com/feed/".to_string(), category: "World".to_string() },
        FeedSource { name: "TechCrunch".to_string(), url: "https://techcrunch.com/feed/".to_string(), category: "Tech".to_string() },
    ]
}

#[derive(Debug, Serialize, Deserialize)]
struct FeedSource {
    name: String,
    url: String,
    category: String,
}

fn process_item(item: &Item, source: &str) -> NewsItem {
    let title = item.title().unwrap_or("No title").to_string();
    let desc = item.description().unwrap_or("").to_string();
    let link = item.link().unwrap_or("").to_string();
    let pub_date = item.pub_date().unwrap_or("").to_string();
    
    let reading_time = estimate_reading_time(&desc);
    let sentiment = analyze_sentiment(&title, &desc);
    let keywords = extract_keywords(&title, &desc);
    let category = infer_category(&title, &desc);
    
    NewsItem {
        title,
        link,
        description: desc,
        pub_date,
        source: source.to_string(),
        category,
        reading_time,
        sentiment,
        keywords,
    }
}